### DISCUSSION TITLE
Workarounds

### DISCUSSION THREAD

**Discussion by rcomer** [SciTools/iris#4149] [Q&A] [answered]
Suppose we put a workaround in Iris to handle a bug in one of its dependencies.  For example

https://github.com/SciTools/iris/blob/83309c32a6c9cfbd4603732a4075bae0f6f07d45/lib/iris/plot.py#L1276-L1280

If the bug is fixed some time later, do we have a policy for removing the workaround, in particular how soon to do that?  This particular bug appears to have been fixed at mpl 3.4.1 (see #4090).  I guess it may not be fair to assume that every user of Iris 3.1 will have the latest mpl version, as it could be pinned in other packages they are using.  So how long is it reasonable to wait?

**Comment by bjlittle** [comment]
@rcomer I suggest that you remove the workaround and we can pin to a minimum version of i.e., `mpl>=3.4.1`

That seem completely reasonable to me to have a minimum bar for dependencies... otherwise workarounds would never be removed from the codebase.

Go for it. If you make a PR, I'd merge it :+1:

**Reply by bjlittle** [reply]
I've just checked and `iris` resolves for `mpl` >=3.4.1 for both `py37` and `py38`, but not `py36` which is pinned back to 3.3.4.

I don't need a great deal of encouragement for us to drop support for `py36`. The sooner we do that the better... `numpy` already have. In fact, there is only 7 months of security patches planned for `py36` before end of life, see https://endoflife.date/python. The latest **stable** version of python is `py39`, so I personally think that we should only ever support at most two versions older than stable, and that would make the minimum bar `py37`.

Let's drop `py36` right now 👍

**Reply by bjlittle** [reply]
Also, the only package that stopping `iris` resolving it's dependencies for `py39` is `pillow` which is pinned back... I'm going to resolve that piece of technical debt (it's also a security risk) then we can roll with `py37`, `py38` and `py39`

**Reply by rcomer** [reply]
Cartopy were looking at adopting a NEP29-style policy, i.e the last 42 months of minor python releases and 24 months of everything else.  42 months would now not include python 3.6 👍
https://github.com/SciTools/cartopy/issues/1518

**Reply by tkknight** [reply]
I have an IEP (Iris Enhancement Proposal) drafted (no PR yet).  I think an IEP for stating our intent on supporting the more recent Python versions would be a way to go.

**Reply by bjlittle** [reply]
@tkknight Nudge acknowledged... let's get that IEP infra-structure up and running :heart:

**Reply by rcomer** [reply]
I'm kinda nervous about the idea of pinning matplotlib to the latest version.  There are a lot of packages out there that interface with mpl, and some may have it pinned to older versions (indeed Iris was in that position till only last summer).  So wouldn't we risk creating conflicts so a user just can't create an environment with the combination of packages they want?

If we remove the workaround and _don't_ pin, the worst that could happen is unexpected results with one specific iplt function.

**Reply by rcomer** [reply, marked as answer]
So I finally opened that PR #4999

**Comment by trexfeathers** [comment]
Is there anything to discuss about how we keep track of such workarounds? There must be a few such things scattered throughout the source code that we have lost awareness of, but maybe that's OK if they're not causing any harm?

**Reply by rcomer** [reply]
We have a test that failed once this particular matplotlib bug was fixed.  I wonder if whoever put the test in did so as a reminder to remove the workaround later?  Although there wasn't a comment to say so.

https://github.com/SciTools/iris/blob/83309c32a6c9cfbd4603732a4075bae0f6f07d45/lib/iris/tests/test_mapping.py#L241-L244

Your question is pertinent to #3790, since I'm proposing a numpy workaround there.

**Reply by rcomer** [reply]
Aside: the PR that added the mpl workaround also added the "swanky new idiff tool" 😆 #109

**Reply by trexfeathers** [reply]
That seems like a good 'self-organising' pattern.

**Reply by rcomer** [reply]
>  I wonder if whoever put the test in did so as a reminder to remove the workaround later

Hmmm probably not actually, as the test was already there, but modified when the workaround was added.
https://github.com/SciTools/iris/pull/109/files#diff-bdbebdfefdcd1f50d7bd045ebf303f65181b7e6534e98c575e907882c78e8e2f

**Comment by pp-mo** [comment]
Discussed again in peloton ..
Perhaps we can label these "TODO:" along with other issues
 --> entry to review these in the release checklist ??
